source/common/buffer/buffer_impl.cc

```diff
@@ -338,55 +299,103 @@ void OwnedImpl::move(Instance& rhs, uint64_t length) {
   other.postProcess();
 }
 
-uint64_t OwnedImpl::reserve(uint64_t length, RawSlice* iovecs, uint64_t num_iovecs) {
-  if (num_iovecs == 0 || length == 0) {
-    return 0;
+Reservation OwnedImpl::reserveApproximately(uint64_t length) {
+  Reservation reservation = Reservation::bufferImplUseOnlyConstruct(*this);
+  if (length == 0) {
+    return reservation;
   }
-  // Check whether there are any empty slices with reservable space at the end of the buffer.
-  size_t first_reservable_slice = slices_.size();
-  while (first_reservable_slice > 0) {
-    if (slices_[first_reservable_slice - 1].reservableSize() == 0) {
-      break;
-    }
-    first_reservable_slice--;
-    if (slices_[first_reservable_slice].dataSize() != 0) {
-      // There is some content in this slice, so anything in front of it is non-reservable.
-      break;
-    }
+
+  // Remove any empty slices at the end.
+  while (!slices_.empty() && slices_.back().dataSize() == 0) {
```

Comment:
Removal of slices with 0 data on reserve could result in the removal of unowned slices with 0 size like the ones added by ServerConnectionImpl::maybeAddSentinelBufferFragment which would break some of the protections against H1 pipelining resulting in infinite sized output buffers.

Possible fix: replace these sentinel slices with buffer drain tracker callbacks.

Reply:
Oh wow, I didn't realize we were doing that! 

I don't think this particular use case will be affected, because we should never be reserving data for read in the http output buffer; only an input buffer should have reserve called on it.